test/search_api/test_search_api_query_filter_factory.py

```diff
@@ -0,0 +1,477 @@
+                ["Instrument 1", "Instrument 1"],
+                ["or", "or"],
+                id="Text operator on instrument",
+            ),
+            pytest.param(
+                {"filter": {"where": {"and": [{"summary": "My Test Summary"}]}}},
+                "documents",
+                1,
+                ["summary"],
+                ["eq"],
+                ["My Test Summary"],
+                ["and"],
+                id="AND with single condition, no specified operator",
+            ),
+            pytest.param(
+                {
+                    "filter": {
+                        "where": {
+                            "and": [
+                                {"summary": "My Test Summary"},
+                                {"title": "Test title"},
+                            ],
+                        },
+                    },
+                },
+                "documents",
+                2,
+                ["summary", "title"],
+                ["eq", "eq"],
+                ["My Test Summary", "Test title"],
+                ["and", "and"],
+                id="AND with multiple conditions, no specified operator",
+            ),
+            pytest.param(
+                {"filter": {"where": {"and": [{"value": {"lt": 50}}]}}},
+                "parameters",
+                1,
+                ["value"],
+                ["lt"],
+                [50],
+                ["and"],
+                id="AND, single condition with operator",
+            ),
+            pytest.param(
+                {
+                    "filter": {
+                        "where": {
+                            "and": [
+                                {"name": {"like": "Test name"}},
+                                {"value": {"gte": 275}},
+                            ],
+                        },
+                    },
+                },
+                "parameters",
+                2,
+                ["name", "value"],
+                ["like", "gte"],
+                ["Test name", 275],
+                ["and", "and"],
+                id="AND, multiple conditions with operator",
+            ),
+            pytest.param(
+                {
+                    "filter": {
+                        "where": {
+                            "and": [
+                                {
+                                    "and": [
+                                        {"name": {"like": "Test name"}},
+                                        {"value": {"gte": 275}},
+                                    ],
+                                },
+                            ],
+                        },
+                    },
+                },
+                "parameters",
+                2,
+                ["name", "value"],
+                ["like", "gte"],
+                ["Test name", 275],
+                ["and", "and"],
+                id="Nested AND, multiple conditions with operator",
+            ),
```

Comment:
I am really confused how a similar OR AND nested query will work.. There is an example one [here](use OR boolean operator for WHERE filters in text operator #259).

With the logic as it is, all WHERE filter objects will end up having an `OR` boolean operator but I don't think this is what we want.

We want this:

`(WHERE Filter 1 AND WHERE Filter 2) OR (WHERE Filter 3 AND WHERE Filter 4)` and not <br>`WHERE Filter 1 OR WHERE Filter 2 OR WHERE Filter 3 OR WHERE Filter 4`.

Text operators inside `AND` will also be affected by this.

Reply:
Why would all where filters default to having an OR boolean operator? [They default to `and`](https://github.com/ral-facilities/datagateway-api/blob/1d24021ce32fe7fa9843de7c438aa6713d4eb44e/datagateway_api/src/search_api/filters.py#L13), they'll only be set to OR if that's passed in at the creation of a `SearchAPIWhereFilter`. Unless you're referring to the nested stuff, maybe we should talk about that over Zoom so we walk through what's going on.

As an aside, I would say the test case which you've attached this comment to is a bit misleading/isn't really nested (my fault) because the top level AND only has one item in it (which contains and AND of 2 items). So you've got `(A AND B) AND` - there's nothing on the right hand side (RHS) of the top-level AND. Do you want me to add a couple of test cases to make this clearer?

I'd probably add the following test cases for nested boolean operators:
- `(A AND B) AND (C AND D)` - in boolean algebra, this is basically just `A AND B AND C AND D`
- `(A AND B) OR (C AND D)`
- `(A OR B) OR (C OR D)` - in boolean algebra, this is basically just `A OR B OR C OR D`
- `(A OR B) AND (C OR D)`
- `(A AND B AND C) OR (C OR D) OR (E AND F)`

Feel free to create the test cases but I'm happy to write them if makes your work easier, just let me know.

Reply:
I thought I updated my comment which I initially wrote in rush but looks like I forgot. The example hyperlink does not also link to the OR AND nested query that I intended to link you to, so apologies for the confusion.

The reason I attached my comment to this test case is because there is an example nested OR AND query that is similar to this case that can be found [here](https://github.com/panosc-eu/search-api/blob/master/doc/dataset-example-queries.md#filter-2). If you look at it, we want to end up having `(A AND B) OR (C AND D)`. With the current implementation, the WHERE filter objects will take the closest logical operator to them and use that as a boolean operator. In the example case, that will be the `and` logical operator but this would mean that we will end up having `A AND B AND C AND D`. With the way the filters are implemented at the moment, I am confused about how to get from that to `(A AND B) OR (C AND D)`. 

Regarding this:
>With the logic as it is, all WHERE filter objects will end up having an `OR` boolean operator...

I was referring to the filter objects in the linked example case and not all. The reason this happens is because the example case is a nested OR AND query so the changes I added [here](https://github.com/ral-facilities/datagateway-api/blob/1d24021ce32fe7fa9843de7c438aa6713d4eb44e/datagateway_api/src/search_api/query_filter_factory.py#L74-L75) modify the boolean operators of the WHERE filters objects to be equal to the logic operators that is the farthest from them which is the `or` logical operator in this case. Now. this is not what we want but there is another reason why I made these changes and I only found about the consequences that it will have with nested logical operator queries after I tried to write a nested OR AND unit test. Essentially the reason was so that we do not have to repeat what we are already doing in the `else` part of the conditional. Instead of creating WHERE filter objects in the `if` part of the conditional, we just create a WHERE filter query and pass that to `get_query_filter` so that the `else` part can deal with the creation of WHERE filter objects. See https://github.com/ral-facilities/datagateway-api/pull/292/commits/a5905148e524fec250c1d381440cca0dc3fab7a7 for more. This works fine for non-nested but does not for nested logical operator queries.

I am happy creating the test cases that you suggested but we need to have a chat on Zoom about the outlined issue.